// minecraft-server-util/dist/structure/UDPSocket.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const dgram_1 = __importDefault(require("dgram"));
const Packet_1 = __importDefault(require("./Packet"));
const TimeoutPromise_1 = __importDefault(require("./TimeoutPromise"));
/**
 * A UDP socket class for reading and writing data to a remote socket.
 * @class
 */
class UDPSocket {
    /**
     * Creates a new UDP socket class from the host and port.
     * @param {string} host The host of the server
     * @param {number} port The port of the server
     * @param {number} timeout The timeout in milliseconds
     * @constructor
     */
    constructor(host, port, timeout) {
        this.host = host;
        this.port = port;
        this.timeout = timeout;
        this.socket = dgram_1.default.createSocket('udp4');
        this.buffer = [];
        this.socket.on('message', (message, info) => {
            this.buffer.push({ info, message });
        });
    }
    /**
     * Reads a packet from the UDP socket.
     * @returns {Promise<Packet>} The packet read from the socket
     * @async
     */
    readPacket() {
        var _a;
        if (this.buffer.length > 0) {
            const packet = new Packet_1.default();
            if (this.buffer.length > 0) {
                const value = this.buffer.shift();
                packet.buffer = (_a = value === null || value === void 0 ? void 0 : value.message) !== null && _a !== void 0 ? _a : Buffer.alloc(0);
            }
            return Promise.resolve(packet);
        }
        const timeoutPromise = new TimeoutPromise_1.default(this.timeout, new Error('Timed out while waiting for server response'));
        const actualPromise = new Promise((resolve, reject) => {
            let read = false;
            const cleanupHandlers = () => {
                this.socket.removeListener('message', messageHandler);
                this.socket.removeListener('error', errorHandler);
                this.socket.removeListener('close', closeHandler);
            };
            const messageHandler = () => {
                var _a;
                if (read)
                    return;
                if (this.buffer.length > 0) {
                    read = true;
                    cleanupHandlers();
                    const packet = new Packet_1.default();
                    if (this.buffer.length > 0) {
                        const value = this.buffer.shift();
                        packet.buffer = (_a = value === null || value === void 0 ? void 0 : value.message) !== null && _a !== void 0 ? _a : Buffer.alloc(0);
                    }
                    resolve(packet);
                }
            };
            const errorHandler = (error) => {
                cleanupHandlers();
                reject(error);
            };
            const closeHandler = () => {
                cleanupHandlers();
                reject(new Error('Socket ended without sending any data back'));
            };
            this.socket.on('message', messageHandler);
            this.socket.on('error', errorHandler);
            this.socket.on('close', closeHandler);
        });
        return Promise.race([
            timeoutPromise.promise,
            actualPromise
        ]);
    }
    /**
     * Writes a packet to the UDP connection.
     * @param {Packet} packet The packet to write to the connection
     * @returns {Promise<void>} A Promise that resolves when it has written the packet
     * @async
     */
    writePacket(packet) {
        return new Promise((resolve, reject) => {
            this.socket.send(Buffer.from(packet.buffer), this.port, this.host, (error) => {
                if (error)
                    return reject(error);
                resolve();
            });
        });
    }
    /**
     * Closes the connection and cleans up data.
     * @returns {Promise<void>} A Promise that resolves when the client has closed
     * @async
     */
    destroy() {
        return new Promise((resolve) => {
            this.socket.removeAllListeners();
            this.socket.close(() => {
                resolve();
            });
        });
    }
}
exports.default = UDPSocket;
